Support Swarmkit Generic Task Runtime

ehazlett opened the pull request:
This adds support for Swarmkit Generic Task Runtimes (https://github.com/docker/swarmkit/pull/2014) to support swarmkit plugins.

Note: the code change is ~ +300.  Most of this PR is the vendor from swarmkit.

This currently has stubs for the plugin support.  I left them in as examples to see how this works.  I can remove and we can add them in a separate PR to keep this PR just on the generic runtime support.

ehazlett committed:
add support for swarmkit generic runtime

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

ehazlett commented:
/cc @tiborvass @anusha-ragunathan @aluzzardi

thaJeztah commented:
there's another SwarmKit bump PR open currently, so we may have to coordinate things :-)

Can you update the bump commit message to include _what_ vendors are bumped for easier finding?

ehazlett commented:
@thaJeztah i update the vendor bump comment with what exact commit is needed from swarmkit.

aaronlehmann commented:
Also note that CI will fail until #31714 is merged.

aaronlehmann commented on daemon/cluster/convert/service.go:
```diff
@@ -56,24 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
+
+	switch t := spec.Task.Runtime.(type) {
+	case *swarmapi.TaskSpec_Container:
+		containerConfig := spec.Task.Runtime.(*swarmapi.TaskSpec_Container).Container
```
`containerConfig := t.Container`

aaronlehmann commented on daemon/cluster/convert/service.go:
```diff
@@ -56,24 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
+
+		taskTemplate.RuntimeData = t.Generic.Payload.Value
+	default:
+		return &types.ServiceSpec{}, fmt.Errorf("error creating service; unsupported runtime %q", t)
```
Maybe `%T` to show the type?

tiborvass replied:
Also it's fine to return `nil` for the servicespec in the error case.

aaronlehmann commented on daemon/cluster/convert/service.go:
```diff
@@ -56,24 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
 		Annotations:  annotationsFromGRPC(spec.Annotations),
-		TaskTemplate: taskSpecFromGRPC(spec.Task),
-		Networks:     serviceNetworks,
-		EndpointSpec: endpointSpecFromGRPC(spec.Endpoint),
```
Removing these fields can't be right.

ehazlett replied:
nope -- rebase fail -- thx!

aaronlehmann commented on daemon/cluster/convert/service.go:
```diff
@@ -124,11 +150,26 @@ func ServiceSpecToGRPC(s types.ServiceSpec) (swarmapi.ServiceSpec, error) {
+			Generic: &swarmapi.GenericRuntimeSpec{
+				Kind: string(types.RuntimePlugin),
+				Payload: &gogotypes.Any{
+					TypeUrl: string(types.RuntimePlugin),
```
`TypeUrl` should be something different from `Kind`, though I'm not sure exactly what its value should be in this case.

aluzzardi replied:
Apparently, `TypeUrl` should be a URL:
https://github.com/gogo/protobuf/blob/master/types/any.go

Some docs here:
https://github.com/gogo/protobuf/blob/master/types/any.pb.go#L109

```
	// * If no scheme is provided, `https` is assumed.
	// * The last segment of the URL's path must represent the fully
	//   qualified name of the type (as in `path/google.protobuf.Duration`).
	//   The name should be in a canonical form (e.g., leading "." is
	//   not accepted).
	// * An HTTP GET on the URL must yield a [google.protobuf.Type][]
	//   value in binary format, or produce an error.
	// * Applications are allowed to cache lookup results based on the
	//   URL, or have them precompiled into a binary to avoid any
	//   lookup. Therefore, binary compatibility needs to be preserved
	//   on changes to types. (Use versioned type names to manage
	//   breaking changes.)
```
Perhaps `type.docker.com/RuntimePlugin`? /cc @stevvooe @crosbymichael

aaronlehmann commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +156,30 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+			}
+			ctlr = c
+		default:
+			return ctlr, fmt.Errorf("unsupported runtime type: %s", t)
```
I think this should be `fmt.Errorf("unsupported runtime type: %T", r)`

aaronlehmann commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +156,30 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+		logrus.WithFields(logrus.Fields{
+			"runtimeUrl": r.Generic.Payload.TypeUrl,
+		}).Debug("custom runtime requested")
+		switch r.Generic.Payload.TypeUrl {
```
This should switch on `r.Generic.Kind`

aaronlehmann commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +156,30 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+		}
+		ctlr = c
+	default:
+		return nil, fmt.Errorf("unsupported runtime: %q", r)
```
`%T`

aaronlehmann commented on daemon/cluster/services.go:
```diff
@@ -102,7 +106,11 @@ func (c *Cluster) GetService(input string) (types.Service, error) {
 		return types.Service{}, err
 	}
-	return convert.ServiceFromGRPC(*service), nil
+	svcs, err := convert.ServiceFromGRPC(*service)
```
How about `svc` or `s`? `svcs` makes me think it's plural.

aaronlehmann commented on daemon/cluster/services.go:
```diff
@@ -123,54 +131,61 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (*apity
+		resp = &apitypes.ServiceCreateResponse{}
+
+		switch serviceSpec.Task.Runtime.(type) {
+		// handle other runtimes herer
```
typo

tiborvass commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +156,30 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+		}).Debug("custom runtime requested")
+		switch r.Generic.Payload.TypeUrl {
+		case string(swarmtypes.RuntimePlugin):
+			c, err := plugin.NewPluginController()
```
i dont think this will compile: it should be `plugin.NewController()`

ehazlett replied:
doh!  thx!

aaronlehmann commented on daemon/cluster/services.go:
```diff
@@ -123,54 +131,61 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (*apity
+
+		switch serviceSpec.Task.Runtime.(type) {
+		// handle other runtimes herer
+		case *swarmapi.TaskSpec_Container:
```
The call above to `serviceSpec.Task.GetContainer()`, and related code, should move into this case body.

For readability, maybe move the body of this case to a separate function... but I'm not sure if there would be too many arguments for this to be practical.

tiborvass commented on api/types/swarm/runtime.go:
```diff
@@ -0,0 +1,11 @@
+
+const (
+	// RuntimeContainer is the container based runtime
+	RuntimeContainer RuntimeType = "com.docker.runtime.container"
```
Were these strings ever used somewhere previously?

aluzzardi replied:
I think right now the one we're using is simply called `container`

@ehazlett There's a `naming.Runtime()` helper to extract the name automatically:
https://github.com/docker/swarmkit/blob/master/api/naming/naming.go#L38

aluzzardi commented on daemon/cluster/convert/service.go:
```diff
@@ -56,24 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
 
+	taskTemplate := taskSpecFromGRPC(spec.Task)
+
+	switch t := spec.Task.Runtime.(type) {
```
Perhaps you could do a `switch runtime := naming.Runtime(spec.Task)` here and switch on the string (`container` vs `com.docker.runtime.plugin`. @aaronlehmann?

If we keep it this way, then rather than switching on `switch t.Generic.Payload.TypeUrl {` we should switch on `t.Generic.Kind`

aaronlehmann replied:
I actually like type switching here because it avoids the type assertion you would need for each value after switching on a string. There's one less opportunity to make mistakes if you just list the types you want to handle instead of having a strings and then needing to match types to those strings.

aaronlehmann replied:
Agreed that switching on `TypeUrl` is wrong below.

aluzzardi commented:
Mostly looks good with a few nits/comments.

This PR should also add a filter in `docker service ls` to select for `runtime=container`

ehazlett commented:
ah yes. This was from a lint fix. Will update!

On Mar 22, 2017 7:33 PM, "Tibor Vass" <notifications@github.com> wrote:

> *@tiborvass* commented on this pull request.
> ------------------------------
>
> In daemon/cluster/executor/container/executor.go
> <https://github.com/docker/docker/pull/32030#discussion_r107558304>:
>
> > @@ -152,9 +156,30 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
>  		return newNetworkAttacherController(e.backend, t, e.secrets)
>  	}
>
> -	ctlr, err := newController(e.backend, t, secrets.Restrict(e.secrets, t))
> -	if err != nil {
> -		return nil, err
> +	var ctlr exec.Controller
> +	switch r := t.Spec.GetRuntime().(type) {
> +	case *api.TaskSpec_Generic:
> +		logrus.WithFields(logrus.Fields{
> +			"runtimeUrl": r.Generic.Payload.TypeUrl,
> +		}).Debug("custom runtime requested")
> +		switch r.Generic.Payload.TypeUrl {
> +		case string(swarmtypes.RuntimePlugin):
> +			c, err := plugin.NewPluginController()
>
> i dont think this will compile: it should be plugin.NewController()
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/32030#pullrequestreview-28538894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAP6Ip6Kc2tx-6BqjYqikV795Y3YWoecks5roa_kgaJpZM4Ml730>
> .
>

ehazlett committed:
filter services by runtime; default to container

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

ehazlett commented:
Addressed comments and added filtering

/cc @aluzzardi @aaronlehmann 

@dnephin the filtering on stacks looked straight forward but if you could take a look i would appreciate it, thx!

aaronlehmann commented on daemon/cluster/convert/service.go:
```diff
@@ -56,22 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
+		taskTemplate.ContainerSpec = containerSpecFromGRPC(containerConfig)
+		taskTemplate.Runtime = types.RuntimeContainer
+	case *swarmapi.TaskSpec_Generic:
+		switch t.Generic.Payload.TypeUrl {
```
This should switch on `t.Generic.Kind`. `t.Generic.Payload.TypeUrl` is a description of the serialization format of the generic payload, not the runtime type. In the future there may be multiple runtimes that share the same `TypeUrl`.

ehazlett replied:
makes sense :+1:

aaronlehmann commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +157,35 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+	case *api.TaskSpec_Generic:
+		logrus.WithFields(logrus.Fields{
+			"kind":       r.Generic.Kind,
+			"runtimeUrl": r.Generic.Payload.TypeUrl,
```
Maybe `type_url` instead of `runtimeUrl`

ehazlett replied:
ah yes :+1:

aaronlehmann commented on daemon/cluster/services.go:
```diff
@@ -43,18 +43,21 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		"id":       true,
+		"label":    true,
+		"mode":     true,
+		"runtimes": true,
```
Why plural when the other keys are singular?

ehazlett replied:
I was matching what was in the swarm `ListServicesRequest_Filters` but it probably makes more sense singular.

ehazlett committed:
updates for review comments

- runtimeUrl -> type_url
- runtimes -> runtime

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

anusha-ragunathan commented on daemon/cluster/convert/service.go:
```diff
@@ -56,22 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
 
+	taskTemplate := taskSpecFromGRPC(spec.Task)
+
+	switch t := spec.Task.Runtime.(type) {
```
Use `spec.GetRuntime().type`, as done in other parts of code?

anusha-ragunathan commented on daemon/cluster/convert/service.go:
```diff
@@ -56,22 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
+
+		taskTemplate.RuntimeData = t.Generic.Payload.Value
+	default:
+		return nil, fmt.Errorf("error creating service; unsupported runtime %T", t)
```
Use `naming.errUnknownRuntime`?

ehazlett replied:
`errUnknownRuntime` is unexported and we wanted to return the type that was specified.

anusha-ragunathan commented on daemon/cluster/convert/service.go:
```diff
@@ -124,11 +152,26 @@ func ServiceSpecToGRPC(s types.ServiceSpec) (swarmapi.ServiceSpec, error) {
+			Generic: &swarmapi.GenericRuntimeSpec{
+				Kind: string(types.RuntimePlugin),
+				Payload: &gogotypes.Any{
+					TypeUrl: string(types.RuntimeURLPlugin),
```
When payload stays generic, then payload url is `com.docker.custom.runtime`. Once it is pluginized, it changes to `types.docker.com/RuntimePlugin`. Can we have it as `com.docker.plugin.runtime` for consistency?

ehazlett replied:
The spec says it needs to be a url such as types.x.x -- I'll update -- good catch :+1:

ehazlett replied:
Actually the more I look at it I'm not following.  If a generic runtime is specified then it sets the kind and payload.  If not, then the `Runtime` is not of the type `TaskSpec_Generic` and it won't have this at all -- it will have a `TaskSpec_Container`.

ehazlett replied:
ah ok i think i found it -- disregard :)

anusha-ragunathan commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -152,9 +157,35 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
+	case *api.TaskSpec_Container:
+		c, err := newController(e.backend, t, secrets.Restrict(e.secrets, t))
+		if err != nil {
+			return nil, err
```
`return ctlr, err`. Same for L188

anusha-ragunathan commented on daemon/cluster/controllers/plugin/controller.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Wait causes the task to wait until returned
+func (p *Controller) Wait(ctx context.Context) error {
```
Is `Wait` necessary for plugin types? What is the behavior from a scheduler point of view?

ehazlett replied:
I believe this is needed to satisfy the interface when called from the backend.  We don't have to use it.

anusha-ragunathan commented on daemon/cluster/convert/service.go:
```diff
@@ -56,22 +64,42 @@ func ServiceFromGRPC(s swarmapi.Service) types.Service {
+			return nil, fmt.Errorf("unknown task runtime type: %s", t.Generic.Payload.TypeUrl)
+		}
+
+		taskTemplate.RuntimeData = t.Generic.Payload.Value
```
Where is this `RuntimeData` byte array allocated?

ehazlett replied:
Sorry I just added a comment that I forgot to mention.  We are planning on using struct tags for protos instead of a byte array.  this is just for the interim to get working functionality.

 /cc @aluzzardi

mavenugo commented:
@ehazlett https://github.com/docker/docker/pull/31714 is merged now. pls rebase.

ehazlett commented:
@mavenugo thanks!

ehazlett committed:
return exec.Controller instead of nil

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

ehazlett committed:
add service convert tests

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

tiborvass commented on daemon/cluster/executor/container/executor.go:
```diff
@@ -154,9 +159,35 @@ func (e *executor) Controller(t *api.Task) (exec.Controller, error) {
-	ctlr, err := newController(e.backend, t, secrets.Restrict(e.secrets, t))
-	if err != nil {
-		return nil, err
+	var ctlr exec.Controller
```
@ehazlett out of curiosity, was it necessary to handle the generic taskspec case in the container executor instead of having a generic executor? Or is the rationale that if we decide to migrate the container spec to generic spec, then the container executor becomes a defacto generic executor? Disregard if this makes no sense.

tiborvass replied:
Nevermind, apparently a node can only have one executor according to `daemon/cluster/noderunner.go` (search for `NewExecutor`)

thaJeztah commented on api/types/swarm/task.go:
```diff
@@ -65,6 +65,11 @@ type TaskSpec struct {
+	Runtime RuntimeType `json:",omitempty"`
+	// TODO (ehazlett): this should be removed and instead
+	// use struct tags (proto) for the runtimes
+	RuntimeData []byte `json:",omitempty"`
```
Don't forget to update the swagger.yml with the new fields; https://github.com/docker/docker/blob/9f9cc221bdb0a919d6ed28bdfaffa9ae2fe3a5fd/api/swagger.yaml#L1894-L2116

thaJeztah replied:
Also, needs a note in the API version-history; https://github.com/docker/docker/blob/master/docs/api/version-history.md#v129-api-changes

thaJeztah commented on cli/command/service/list.go:
```diff
@@ -45,7 +45,9 @@ func runList(dockerCli *command.DockerCli, opts listOptions) error {
 
-	services, err := client.ServiceList(ctx, types.ServiceListOptions{Filters: opts.filter.Value()})
+	serviceFilters := opts.filter.Value()
+	serviceFilters.Add("runtime", string(swarm.RuntimeContainer))
```
IIUC, this adds a new filter option to `docker service ls`, and the API for that command? If so, please add the new filter to the CLI docs; https://github.com/docker/docker/blob/master/docs/reference/commandline/service_ls.md#filtering, and the API history https://github.com/docker/docker/blob/master/docs/api/version-history.md#v129-api-changes

thaJeztah commented on cli/command/service/ps.go:
```diff
@@ -58,8 +59,11 @@ func runPS(dockerCli *command.DockerCli, opts psOptions) error {
 		serviceIDFilter.Add("id", service)
+		serviceIDFilter.Add("runtime", string(swarmtypes.RuntimeContainer))
 		serviceNameFilter.Add("name", service)
+		serviceNameFilter.Add("runtime", string(swarmtypes.RuntimeContainer))
```
This is adding the same filter _twice_?

ehazlett replied:
I don't believe so.  `Add` adds to a map and it's setting the key `runtime` in addition to `name`.

thaJeztah commented on cli/command/service/ps.go:
```diff
@@ -58,8 +59,11 @@ func runPS(dockerCli *command.DockerCli, opts psOptions) error {
 	for _, service := range opts.services {
+		// default to container runtime
 		serviceIDFilter.Add("id", service)
+		serviceIDFilter.Add("runtime", string(swarmtypes.RuntimeContainer))
```
IIUC, this adds a new filter option to `docker service ls`, and the API for that command? If so, please add the new filter to the CLI docs; https://github.com/docker/docker/blob/master/docs/reference/commandline/service_ps.md#filtering, and the API history https://github.com/docker/docker/blob/master/docs/api/version-history.md#v129-api-changes

ehazlett replied:
These are a bit different.  This cannot be called from the API as these are intended to be used internally for filtering.  Should we doc these now or wait until we support the custom filtering based upon runtime type?

thaJeztah commented on cli/command/stack/common.go:
```diff
@@ -36,7 +42,7 @@ func getServices(
 	return apiclient.ServiceList(
 		ctx,
-		types.ServiceListOptions{Filters: getStackFilter(namespace)})
+		types.ServiceListOptions{Filters: getServiceFilter(namespace)})
```
Same here; is this adding new filter options to the `docker stack ls` / `docker stack services` and `docker stack ps` commands?

ehazlett replied:
(same comment as above )

These are intended to be used internally for filtering. Should we doc these now or wait until we support the custom filtering based upon runtime type?

anusha-ragunathan replied:
I think we should wait until the filtering is exposed externally.

thaJeztah reviewed (changes requested):
left some comments / questions :smile:

cpuguy83 commented on api/types/swarm/runtime.go:
```diff
@@ -0,0 +1,19 @@
+	RuntimePlugin RuntimeType = "plugin"
+
+	// RuntimeURLContainer is the proto url for the container type
+	RuntimeURLContainer RuntimeURL = "types.docker.com/RuntimeContainer"
```
wouldn't this typically be reverse? `com.docker.types`?

ehazlett replied:
It looks like it's supposed to be a Url (https://github.com/gogo/protobuf/blob/master/types/any.pb.go#L109 -- example https://github.com/gogo/protobuf/blob/master/types/any.pb.go#L104).

aluzzardi replied:
I don't think this is actually used, yet?

anusha-ragunathan commented:
LGTM

tiborvass commented:
Needs update in swagger.yml as thajeztah mentioned.

ehazlett commented:
@tiborvass I replied (https://github.com/docker/docker/pull/32030#discussion_r109938112) -- I wasn't sure if we should add them as they are internal only and cannot be modified.

anusha-ragunathan commented:
@ehazlett : swagger.yaml needs to be updated with the fields added to `TaskSpec`, Runtime and RuntimeData.

ehazlett committed:
updated swagger with new fields for task spec

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

ehazlett commented:
@anusha-ragunathan updated swagger /cc @thaJeztah 
I didn't update the example as the fields are not user accessible.

tiborvass commented:
Let's take this for 17.06, and not 17.05.

tiborvass commented:
LGTM

tiborvass commented:
@anusha-ragunathan @thaJeztah FYI i removed the impact labels.

aluzzardi commented on api/swagger.yaml:
```diff
@@ -2096,6 +2096,12 @@ definitions:
+        description: "Runtime is the type of runtime specified for the task executor."
+        type: "string"
+      RuntimeData:
+        description: "RuntimeData is the payload sent to be used with the runtime for the executor."
```
do we want to expose this to the user? Rather than a well-formed object (however docker decides to do so)?

aluzzardi replied:
I don't think it should be exposed - it's internal data, however we decide to represent that.

The only user is going to be internal code, so I don't think we need to expose that at all

tiborvass replied:
@aluzzardi we'll change this before 17.06, there's a TODO. Just trying to minimize rebase churn.

cpuguy83 replied:
Looks like this was never cleaned up?

thaJeztah replied:
@cpuguy83 can you open an issue for it so that we fix before GA? (hopefully)

ehazlett replied:
Opened https://github.com/moby/moby/pull/33308 to address this.
